### PR TITLE
LP-2703 Add References section on application page

### PR DESCRIPTION
**Pull request by harris-irfan**
[Ticket](https://philanthropyu.atlassian.net/browse/LP-2703)
[adg-edx-theme PR](https://github.com/OmnipreneurshipAcademy/adg-edx-theme/pull/305)

### CODE REVIEW

**Comment by omnipreneurshipA** [comment]
## Diff Coverage Report - ADG
Check-out complete **[report](https://app.circleci.com/pipelines/github/OmnipreneurshipAcademy/edx-platform/1977/workflows/workflows_id/jobs/1977/artifacts)**. Coverage !!! :rocket:
```
-------------
Diff Coverage
Diff: origin/develop...HEAD, staged and unstaged changes
-------------
openedx/adg/lms/applications/admin.py (100%)
openedx/adg/lms/applications/api/serializers.py (100%)
openedx/adg/lms/applications/api/urls.py (100%)
openedx/adg/lms/applications/api/views.py (100%)
openedx/adg/lms/applications/constants.py (100%)
openedx/adg/lms/applications/models.py (91.7%): Missing lines 451
openedx/adg/lms/applications/tests/factories.py (100%)
openedx/adg/lms/applications/tests/test_api/test_serializers.py (100%)
openedx/adg/lms/applications/tests/test_api/test_views.py (100%)
openedx/adg/lms/applications/tests/test_views.py (100%)
openedx/adg/lms/applications/views.py (75.0%): Missing lines 234
-------------
Total:   68 lines
Missing: 2 lines
Coverage: 97%
-------------
```

File: openedx/adg/lms/applications/constants.py
Comment (MunamMubashir-arbisoft) [inline openedx/adg/lms/applications/constants.py:78]: Generally, we write error message of this kind in the form, 

> This field cannot exceed...

Perhaps we can phrase this message in that way as well for consistency.

File: openedx/adg/lms/applications/views.py
Comment (MunamMubashir-arbisoft) [inline openedx/adg/lms/applications/views.py:294]: In my opinion, EducationAndExperience completely explains all that the view entails. 
The view or page basically contains Education, Work Experience, Background Question (about experience) and References (about experience and education). Therefore, we should not rename this.

Secondly, when we are merging the two tickets, this might create a hassle.

File: openedx/adg/lms/applications/tests/test_api/test_views.py
Comment (MunamMubashir-arbisoft) [inline openedx/adg/lms/applications/tests/test_api/test_views.py:364]: We could also test that the required attributes are provided or not. And if they are, is the validation working as it should.

**Review by MunamMubashir-arbisoft — CHANGES_REQUESTED** [review]
Great work bro! 
Just a few minor suggestions.

File: openedx/adg/lms/applications/constants.py
Comment (harris-irfan) [reply openedx/adg/lms/applications/constants.py:78]: Actually this error message is not related to the field. It is a non-field error message that is displayed when a client attempts to add more than 2 references against a user application.

File: openedx/adg/lms/applications/tests/test_api/test_views.py
Comment (harris-irfan) [reply openedx/adg/lms/applications/tests/test_api/test_views.py:364]: Well, in this case, everything is handled by the viewset provided by Django. We are using ModelViewset. If the required attributes are not provided, a validation error would be raised by the serializer, and I'm not customizing the serializer, just using the ModelSerializer. The only thing I'm customizing in the entire API, is the create action to only keep a check at the max references. So only that needs to be tested.

**Comment by omnipreneurshipA** [comment]
## Diff Coverage Report - ADG
Check-out complete **[report](https://app.circleci.com/pipelines/github/OmnipreneurshipAcademy/edx-platform/1999/workflows/workflows_id/jobs/1999/artifacts)**. Coverage !!! :rocket:
```
-------------
Diff Coverage
Diff: origin/develop...HEAD, staged and unstaged changes
-------------
openedx/adg/lms/applications/admin.py (100%)
openedx/adg/lms/applications/api/serializers.py (100%)
openedx/adg/lms/applications/api/tests/test_serializers.py (100%)
openedx/adg/lms/applications/api/tests/test_views.py (100%)
openedx/adg/lms/applications/api/urls.py (100%)
openedx/adg/lms/applications/api/views.py (100%)
openedx/adg/lms/applications/constants.py (100%)
openedx/adg/lms/applications/models.py (91.7%): Missing lines 451
openedx/adg/lms/applications/tests/factories.py (100%)
openedx/adg/lms/branding_extension/constants.py (100%)
openedx/adg/lms/branding_extension/helpers.py (100%)
openedx/adg/lms/branding_extension/tests/helpers_tests.py (100%)
-------------
Total:   64 lines
Missing: 1 line
Coverage: 98%
-------------
```

**Review by asadali145 — CHANGES_REQUESTED** [review]
Nice work, I have added a few suggestions, have a look please

File: openedx/adg/lms/applications/constants.py
Comment (harris-irfan) [reply openedx/adg/lms/applications/constants.py:78]: A: Making it translatable doesn't cost us anything, so I think It's better if we do. I understand that the error message would be displayed if someone is calling it directly via API, and it is not exactly user facing, since we do not allow the user from the front end to add more than 2 references. The translation could be of potential help for the future, but I can remove it if you want me to.

File: openedx/adg/lms/applications/api/serializers.py
Comment (harris-irfan) [reply openedx/adg/lms/applications/api/serializers.py:114]: Taken from DRF [documentation](https://www.django-rest-framework.org/api-guide/serializers/):

"It is strongly recommended that you explicitly set all fields that should be serialized using the fields attribute. This will make it less likely to result in unintentionally exposing data when your models change."

File: openedx/adg/lms/applications/api/views.py
Comment (harris-irfan) [reply openedx/adg/lms/applications/api/views.py:158]: You are right, I missed it while copying it from Education and Work Experience viewsets' docstrings.

File: openedx/adg/lms/applications/api/views.py
Comment (harris-irfan) [reply openedx/adg/lms/applications/api/views.py:198]: This validation is intentionally added here. If we add a `validate_user_application` method in the serializer class, it will perform the validation for PUT requests as well, and won't allow updating a reference of a user application if two references already exist against that user application. We specifically want the validation to be applied against the `create` action, i.e. a POST request.

File: openedx/adg/lms/applications/models.py
Comment (harris-irfan) [reply openedx/adg/lms/applications/models.py:439]: I don't there is any need to, unless there is a specific requirement from the client. We are using this `max_length=255` as default in many other cases, for example the `job_position_title` field in `WorkExperience` model, which serves an exactly similar purpose.

**Comment by omnipreneurshipA** [comment]
## Diff Coverage Report - ADG
Check-out complete **[report](https://app.circleci.com/pipelines/github/OmnipreneurshipAcademy/edx-platform/2007/workflows/workflows_id/jobs/2007/artifacts)**. Coverage !!! :rocket:
```
-------------
Diff Coverage
Diff: origin/develop...HEAD, staged and unstaged changes
-------------
openedx/adg/lms/applications/admin.py (100%)
openedx/adg/lms/applications/api/serializers.py (100%)
openedx/adg/lms/applications/api/tests/test_serializers.py (100%)
openedx/adg/lms/applications/api/tests/test_views.py (100%)
openedx/adg/lms/applications/api/urls.py (100%)
openedx/adg/lms/applications/api/views.py (100%)
openedx/adg/lms/applications/constants.py (100%)
openedx/adg/lms/applications/models.py (91.7%): Missing lines 451
openedx/adg/lms/applications/tests/factories.py (100%)
-------------
Total:   57 lines
Missing: 1 line
Coverage: 98%
-------------
```

File: openedx/adg/lms/applications/constants.py
Comment (ahmed-faraz46) [reply openedx/adg/lms/applications/constants.py:78]: If we would go with a validator, let's move the error inside their too because not something we could reuse at multiple places and our non-field error message are also in the validators. Also is there any specific reason behind lazy formatting it? like there is a constant we are using in there and it would be ready even at application load time.

File: openedx/adg/lms/applications/views.py
Comment (ahmed-faraz46) [inline openedx/adg/lms/applications/views.py:294]: Should we reconsider this name, as we have added a new major section on this page? I couldn't think of a generic name which could cover all these things, if we could, that would be great otherwise we can keep it something as simple as `EducationExperienceAndReferenceView`. Same goes for the template name

**Review by ahmed-faraz46 — CHANGES_REQUESTED** [review]
Well thought and well researched implementation! kudos for going an extra mile and moving apis in a separate directory! Here a few minor comments/suggestions to look upon. Rest looks great!

File: openedx/adg/lms/applications/api/urls.py
Comment (harris-irfan) [reply openedx/adg/lms/applications/api/urls.py:6]: Didn't we have a team-wide consensus to not necessarily use absolute paths for modules in the same directory, but definitely to use absolute paths for modules placed anywhere else? I think this practice of not using absolute paths for same module imports from the same directory is widely used all over the project.

File: openedx/adg/lms/applications/constants.py
Comment (harris-irfan) [reply openedx/adg/lms/applications/constants.py:78]: @ahmed-faraz46 I am now using a validator, but I'm still keeping the error message in the `constants.py` file since I'm also using it in the unit test. I've removed the lazy formatting, thanks for pointing it out.

File: openedx/adg/lms/applications/views.py
Comment (harris-irfan) [reply openedx/adg/lms/applications/views.py]: @ahmed-faraz46 the reason behind not changing the name of the view and templates ^

File: openedx/adg/lms/applications/views.py
Comment (harris-irfan) [reply openedx/adg/lms/applications/views.py:294]: tagged u in a comment

**Comment by omnipreneurshipA** [comment]
## Diff Coverage Report - ADG
Check-out complete **[report](https://app.circleci.com/pipelines/github/OmnipreneurshipAcademy/edx-platform/2010/workflows/workflows_id/jobs/2010/artifacts)**. Coverage !!! :rocket:
```
-------------
Diff Coverage
Diff: origin/develop...HEAD, staged and unstaged changes
-------------
openedx/adg/lms/applications/admin.py (100%)
openedx/adg/lms/applications/api/serializers.py (92.9%): Missing lines 119
openedx/adg/lms/applications/api/tests/test_serializers.py (100%)
openedx/adg/lms/applications/api/tests/test_views.py (100%)
openedx/adg/lms/applications/api/urls.py (100%)
openedx/adg/lms/applications/api/views.py (100%)
openedx/adg/lms/applications/constants.py (100%)
openedx/adg/lms/applications/models.py (91.7%): Missing lines 451
openedx/adg/lms/applications/tests/factories.py (100%)
-------------
Total:   58 lines
Missing: 2 lines
Coverage: 96%
-------------
```

File: openedx/adg/lms/applications/api/urls.py
Comment (ahmed-faraz46) [reply openedx/adg/lms/applications/api/urls.py:6]: Thanks for the reference, Actually highlighted as I thought we are not consistent within this PR, but now I've noticed that those imports which are changed to absolute are due to create of a subdirectory.

File: openedx/adg/lms/applications/api/serializers.py
Comment (ahmed-faraz46) [inline openedx/adg/lms/applications/api/serializers.py:119]: Let's please describe as add a comment here, or in the doc string that this is to identify if the call is for creation or any other operations

**Review by ahmed-faraz46 — APPROVED** [review]
Looks good to me! Approving with an other minor suggestion

File: openedx/adg/lms/applications/api/serializers.py
Comment (harris-irfan) [reply openedx/adg/lms/applications/api/serializers.py:119]: I thought about adding a comment here, but don't you think that it is quite obvious by looking at the code that if an instance is provided to the serializer, it means that it is an update or delete operation, and if an instance is not provided, it is of course a new creation (POST) operation. We also have adding an explanatory docstring, so I feel an extra comment is not needed. As per clean code guidelines, we should avoid code comments as much as possible and strive towards making our code self-explanatory. I feel this is explanatory enough, as we use this instance approach in many other places as well, like forms, etc.

But of course, I'm no the right person to judge this, since I've written this code. So if you still feel that we require a comment, I'll be more than happy to add one.

File: openedx/adg/lms/applications/api/tests/test_serializers.py
Comment (MunamMubashir-arbisoft) [inline openedx/adg/lms/applications/api/tests/test_serializers.py:172]: Let's please save rewriting the last two asserts by parametrizing this test and receive the `is_valid`, `errors` (if applicable) for different cases.

File: openedx/adg/lms/applications/views.py
Comment (MunamMubashir-arbisoft) [inline openedx/adg/lms/applications/views.py:296]: Are we calling the second page of the written application, _My Experience_?
Better to name this Education & Experience page (as it was called before) for now, till we've changed it in the other places too. WDYT?

**Review by MunamMubashir-arbisoft — CHANGES_REQUESTED** [review]
Great work @harris-irfan, very clean. 👏🏼 
I have a couple small suggestions only, please have a look.

File: openedx/adg/lms/applications/api/tests/test_serializers.py
Comment (harris-irfan) [reply openedx/adg/lms/applications/api/tests/test_serializers.py:172]: These are not two separate test cases, but just one test case. The last two asserts are dependent upon the rest of the code in the test function. The first set of assert statements check the boundary condition, i.e. let us say if max number of references are 2, and there is one reference present, the serializer should validate that reference. Then a new reference is created, so that now there are two references present, i.e. equal to the max limit. The second set of asserts checks that the serializer does not validate a third reference.

If we were parameterize it like you are suggesting, then it would result in unnecessary extra computation. The valid test will create `max - 1` references and then execute the asserts. The invalid test will again create `max-1` references, and another extra reference, and then execute the asserts.

File: openedx/adg/lms/applications/views.py
Comment (harris-irfan) [reply openedx/adg/lms/applications/views.py:296]: We had not changed the names of the template and view to avoid future code conflicts and agreed that the existing names would suffice. However, this is a docstring, and has no potential implications on rest of the code. The reason behind calling the 2nd page of the application "My Experience" page is that it is the title of the second page as per the designs, so it makes sense to call it that. "Education & Experience page" is not accommodating enough since it does not directly cater to the background question and the references (despite the fact that background question and references are somewhat related to education and experience). Ahmed was suggesting to change the name of the template and view as well, but I didn't do that owing to the discussion we had of not changing view and file names.

File: openedx/adg/lms/applications/api/tests/test_serializers.py
Comment (MunamMubashir-arbisoft) [reply openedx/adg/lms/applications/api/tests/test_serializers.py:172]: It won't really be that much extra computation, and it the recommended way of testing things. We can perhaps take the number of references to be created as a parameter, and then bulk_create that number of references, that would address the extra computation. Secondly, as far as I can tell the first assert and the second assert are not directly dependent on each other i.e either an application will have one reference or two references, it can't be both.

I would still suggest parametrizing this, but I leave the last decision on you.

File: openedx/adg/lms/applications/views.py
Comment (MunamMubashir-arbisoft) [reply openedx/adg/lms/applications/views.py:296]: That makes sense! Thank you for filling me in.

**Review by MunamMubashir-arbisoft — APPROVED** [review]

File: openedx/adg/lms/applications/api/serializers.py
Comment (harris-irfan) [reply openedx/adg/lms/applications/api/serializers.py:119]: @MunamMubashir-arbisoft as a reviewer, would you like to share your thoughts on this?

Comment (MunamMubashir-arbisoft) [reply openedx/adg/lms/applications/api/serializers.py:119]: I think we should add a line in the docstring to better communicate the method's function (no pun intended). A single line to explain it similar to how you explained it Ahmed bhai.

File: openedx/adg/lms/applications/api/tests/test_serializers.py
Comment (harris-irfan) [reply openedx/adg/lms/applications/api/tests/test_serializers.py:172]: Do you think that setting default value as `undefined` will be a better approach to this according to programatic conventions. According to that, your concern of not assigning null to variables other than those which are meant to hold objects would also be addressed.

File: openedx/adg/lms/applications/api/views.py
Comment (MunamMubashir-arbisoft) [inline openedx/adg/lms/applications/api/views.py:188]: Kindly use `Reference.objects.all().select_related('user_application')` as it will save us an extra query for every single reference object returned.

**Review by MunamMubashir-arbisoft — CHANGES_REQUESTED** [review]
@harris-irfan Sorry I missed this before. Kindly, make this extra change before merging the PR.

File: openedx/adg/lms/applications/api/views.py
Comment (harris-irfan) [reply openedx/adg/lms/applications/api/views.py:188]: I'm not sure if we really need this. No where am I accessing the related object i.e. `user_application` via the reference. Had I been doing something like that:

```
applications = set()

for reference in queryset:
    applications.add(reference.user_application)
```

only then prefetching the related object of user application in advance with the queryset would have been helpful, because it would save extra db operations each time user_application is accessed.

With the current operations, none of the default actions require accessing the user_application object. For example, we are using list, create, update and delete operations via this Reference API. Listing a record does not list its related objects, it just lists the id of the related object, hence no separate call to the deb is made to access the user_application. Create action does not use the queryset to begin with. We are not allowing nested updates, i.e. updating the user_application via the reference API, only name, email, etc. are updated via the reference API and in that case as well, the user_application is not accessed. In case of delete operation, there is also no access of the related objects.

So I think using `select_related` to fetch `user_application` in advance would not save us an operation but result in extra needless operations as underneath it uses the SQL join and fetches the related instance, and join is a heavy process. Plus, we are using the education and work experience API in the same manner, and we are not using `select_related` in their querysets either.

If I am missing something, please fill me in

File: openedx/adg/lms/applications/api/tests/test_serializers.py
Comment (harris-irfan) [reply openedx/adg/lms/applications/api/tests/test_serializers.py:172]: I've accommodated the suggestion. Thanks for going the extra mile and convince me for it. I believe you are right.

**Comment by omnipreneurshipA** [comment]
## Diff Coverage Report - ADG
Check-out complete **[report](https://app.circleci.com/pipelines/github/OmnipreneurshipAcademy/edx-platform/2015/workflows/workflows_id/jobs/2015/artifacts)**. Coverage !!! :rocket:
```
-------------
Diff Coverage
Diff: origin/develop...HEAD, staged and unstaged changes
-------------
openedx/adg/lms/applications/admin.py (100%)
openedx/adg/lms/applications/api/serializers.py (100%)
openedx/adg/lms/applications/api/tests/test_serializers.py (100%)
openedx/adg/lms/applications/api/tests/test_views.py (100%)
openedx/adg/lms/applications/api/urls.py (100%)
openedx/adg/lms/applications/api/views.py (100%)
openedx/adg/lms/applications/constants.py (100%)
openedx/adg/lms/applications/models.py (91.7%): Missing lines 451
openedx/adg/lms/applications/tests/factories.py (100%)
-------------
Total:   74 lines
Missing: 1 line
Coverage: 98%
-------------
```

File: openedx/adg/lms/applications/api/serializers.py
Comment (ahmed-faraz46) [reply openedx/adg/lms/applications/api/serializers.py:122]: @harris-irfan actually it is not that self explanatory at least readers would have to think about it for some time before they get it, if we really want to make it self explanatory without adding a comment, then we can write a private method here, `def is_create_request` and check is instance available or not in that method and here we will simply add a check

```
if is_create_request() and Reference.objects.filter(user_application=value).count() == MAX_NUMBER_OF_REFERENCES:         
           raise serializers.ValidationError(MAX_REFERENCE_ERROR_MSG)
```
This long line could also be improved by defining a variable above.

Comment (harris-irfan) [reply openedx/adg/lms/applications/api/serializers.py:122]: @ahmed-faraz46 I've updated the docstring. If it suffices, good, otherwise I'll do as you mentioned.

**Comment by omnipreneurshipA** [comment]
## Diff Coverage Report - ADG
Check-out complete **[report](https://app.circleci.com/pipelines/github/OmnipreneurshipAcademy/edx-platform/2018/workflows/workflows_id/jobs/2018/artifacts)**. Coverage !!! :rocket:
```
-------------
Diff Coverage
Diff: origin/develop...HEAD, staged and unstaged changes
-------------
openedx/adg/lms/applications/admin.py (100%)
openedx/adg/lms/applications/api/serializers.py (100%)
openedx/adg/lms/applications/api/tests/test_serializers.py (100%)
openedx/adg/lms/applications/api/tests/test_views.py (100%)
openedx/adg/lms/applications/api/urls.py (100%)
openedx/adg/lms/applications/api/views.py (100%)
openedx/adg/lms/applications/constants.py (100%)
openedx/adg/lms/applications/models.py (91.7%): Missing lines 451
openedx/adg/lms/applications/tests/factories.py (100%)
-------------
Total:   76 lines
Missing: 1 line
Coverage: 98%
-------------
```

File: openedx/adg/lms/applications/api/views.py
Comment (harris-irfan) [reply openedx/adg/lms/applications/api/views.py:188]: @ahmed-faraz46

Comment (ahmed-faraz46) [reply openedx/adg/lms/applications/api/views.py:188]: I guess the explanation is quite logical, and Munam is also scheduling a session around select_related, prefetch_related and some other optimization techniques, we would cover this case in that session as well and then if it would be required we will change this too under a separate ticket with other optimizations.

Comment (MunamMubashir-arbisoft) [reply openedx/adg/lms/applications/api/views.py:188]: Sorry for the delayed response, but I assumed we're using the user_application to link the reference to the application. And it would definitely be accessed at that time. However, that might not be the case in this exact viewset as we are only doing a get but not displaying any user_application . Thank you for the explanation, Harris.

**Review by MunamMubashir-arbisoft — APPROVED** [review]

**Pull request merged**
